File: core/scoreboard.go
package core

import (
	"encoding/json"
	"os"
	"sort"
	"sync"
	"time"
)

// NetworkID identifies the network the phone is on right now. What works on
// home wifi is often not what works on mobile data, and an engine that died on
// one carrier says nothing about the other, so every score is kept per network.
//
// It is a coarse fingerprint on purpose: no SSID, no IP, nothing that would be
// worth handing to anyone who got hold of the phone.
type NetworkID string

type record struct {
	Wins     int           `json:"wins"`
	Losses   int           `json:"losses"`
	Latency  time.Duration `json:"latency"` // moving average of time to first working probe
	LastGood time.Time     `json:"last_good"`
}

type Scoreboard struct {
	path string

	mu     sync.Mutex
	Nets   map[NetworkID]map[string]*record `json:"nets"`
	Boot   map[string]bool                  `json:"bootstrapped"`
	dirty  bool
}

func LoadScoreboard(path string) *Scoreboard {
	b := &Scoreboard{
		path: path,
		Nets: map[NetworkID]map[string]*record{},
		Boot: map[string]bool{},
	}
	if data, err := os.ReadFile(path); err == nil {
		_ = json.Unmarshal(data, b)
		if b.Nets == nil {
			b.Nets = map[NetworkID]map[string]*record{}
		}
		if b.Boot == nil {
			b.Boot = map[string]bool{}
		}
	}
	return b
}

func (b *Scoreboard) Save() error {
	b.mu.Lock()
	defer b.mu.Unlock()
	if !b.dirty {
		return nil
	}
	data, err := json.Marshal(b)
	if err != nil {
		return err
	}
	b.dirty = false
	return os.WriteFile(b.path, data, 0o600)
}

func (b *Scoreboard) Record(net NetworkID, engine string, ok bool, took time.Duration) {
	b.mu.Lock()
	defer b.mu.Unlock()

	byEngine, seen := b.Nets[net]
	if !seen {
		byEngine = map[string]*record{}
		b.Nets[net] = byEngine
	}
	rec, seen := byEngine[engine]
	if !seen {
		rec = &record{}
		byEngine[engine] = rec
	}

	if ok {
		rec.Wins++
		rec.LastGood = time.Now()
		if rec.Latency == 0 {
			rec.Latency = took
		} else {
			// Weighted toward history so one slow night does not bury an
			// engine that is usually the fastest one here.
			rec.Latency = (rec.Latency*3 + took) / 4
		}
	} else {
		rec.Losses++
	}
	b.dirty = true
}

// Order decides who goes first. Anything that worked on this network recently
// leads, sorted by how quickly it came up. Engines with no history here go
// next, because an unknown is worth more than a known failure. Whatever has
// been failing here goes last, but still goes — networks change.
func (b *Scoreboard) Order(net NetworkID, engines []Engine) []Engine {
	b.mu.Lock()
	byEngine := b.Nets[net]
	scores := map[string]*record{}
	for name, rec := range byEngine {
		copied := *rec
		scores[name] = &copied
	}
	b.mu.Unlock()

	type ranked struct {
		engine Engine
		tier   int
		speed  time.Duration
	}

	fresh := time.Now().Add(-72 * time.Hour)
	list := make([]ranked, 0, len(engines))

	for _, e := range engines {
		rec, seen := scores[e.Name()]
		switch {
		case seen && rec.LastGood.After(fresh):
			list = append(list, ranked{e, 0, rec.Latency})
		case !seen:
			list = append(list, ranked{e, 1, 0})
		default:
			list = append(list, ranked{e, 2, 0})
		}
	}

	sort.SliceStable(list, func(i, j int) bool {
		if list[i].tier != list[j].tier {
			return list[i].tier < list[j].tier
		}
		return list[i].speed < list[j].speed
	})

	out := make([]Engine, len(list))
	for i, r := range list {
		out[i] = r.engine
	}
	return out
}

// Bootstrapped reports whether an engine has what it needs to start cold.
func (b *Scoreboard) Bootstrapped(engine string) bool {
	b.mu.Lock()
	defer b.mu.Unlock()
	return b.Boot[engine]
}

func (b *Scoreboard) MarkBootstrapped(engine string) {
	b.mu.Lock()
	b.Boot[engine] = true
	b.dirty = true
	b.mu.Unlock()
}

File: core/src/main/kotlin/xyz/jmc/gozar/core/Scoreboard.kt
package xyz.jmc.gozar.core

/**
 * Which network the phone is on right now.
 *
 * Deliberately coarse: no SSID, no IP, nothing worth finding on a phone that
 * ended up in the wrong hands. "mobile:MCI" and "wifi" is as specific as it
 * gets, and that is enough — what works on a carrier is rarely what works on
 * home wifi, and that is the only distinction the scoreboard needs.
 */
@JvmInline
value class NetworkId(val value: String) {
    companion object {
        val UNKNOWN = NetworkId("unknown")
        fun wifi() = NetworkId("wifi")
        fun mobile(carrier: String?) = NetworkId("mobile:${carrier.orEmpty().ifEmpty { "?" }}")
    }
}

data class EngineRecord(
    val wins: Int = 0,
    val losses: Int = 0,
    /** Moving average of how long it took to come up and pass a probe. */
    val avgMs: Long = 0,
    val lastGoodAt: Long = 0,
)

/** Somewhere to keep the scoreboard between launches. Android hands in a file. */
interface ScoreStore {
    fun load(): Map<String, Map<String, EngineRecord>>
    fun save(data: Map<String, Map<String, EngineRecord>>, bootstrapped: Set<String>)
    fun loadBootstrapped(): Set<String>
}

/** A store that forgets everything. Useful for tests and first runs. */
class MemoryStore : ScoreStore {
    private var data: Map<String, Map<String, EngineRecord>> = emptyMap()
    private var boot: Set<String> = emptySet()
    override fun load() = data
    override fun loadBootstrapped() = boot
    override fun save(data: Map<String, Map<String, EngineRecord>>, bootstrapped: Set<String>) {
        this.data = data
        this.boot = bootstrapped
    }
}

/**
 * Remembers what worked where.
 *
 * This is what makes the second launch fast. The first one is a real race; every
 * one after it starts with whatever won here last time and usually finishes
 * before the second engine has even been started.
 */
class Scoreboard(
    private val store: ScoreStore,
    private val clock: () -> Long = System::currentTimeMillis,
) {
    private val nets: MutableMap<String, MutableMap<String, EngineRecord>> =
        store.load().mapValues { it.value.toMutableMap() }.toMutableMap()
    private val bootstrapped: MutableSet<String> = store.loadBootstrapped().toMutableSet()

    @Synchronized
    fun record(network: NetworkId, engine: String, ok: Boolean, tookMs: Long) {
        val byEngine = nets.getOrPut(network.value) { mutableMapOf() }
        val current = byEngine[engine] ?: EngineRecord()

        byEngine[engine] = if (ok) {
            current.copy(
                wins = current.wins + 1,
                lastGoodAt = clock(),
                // Weighted toward history, so one bad night does not bury the
                // engine that is usually fastest here.
                avgMs = if (current.avgMs == 0L) tookMs else (current.avgMs * 3 + tookMs) / 4,
            )
        } else {
            current.copy(losses = current.losses + 1)
        }
        persist()
    }

    /**
     * Decides who goes off the line first.
     *
     * Anything that worked here recently leads, fastest first. Engines with no
     * history here come next, because an unknown is worth more than a known
     * failure. What has been failing here goes last — but it still goes, because
     * networks change and yesterday's dead engine is sometimes today's only one.
     */
    @Synchronized
    fun order(network: NetworkId, engines: List<Engine>): List<Engine> {
        val scores = nets[network.value].orEmpty()
        val fresh = clock() - Racer.RECENT_WIN_MS

        return engines.sortedWith(
            compareBy(
                { engine ->
                    val rec = scores[engine.name]
                    when {
                        rec != null && rec.lastGoodAt > fresh -> 0
                        rec == null -> 1
                        else -> 2
                    }
                },
                { engine -> scores[engine.name]?.avgMs ?: Long.MAX_VALUE },
            )
        )
    }

    @Synchronized
    fun isBootstrapped(engine: String) = engine in bootstrapped

    @Synchronized
    fun markBootstrapped(engine: String) {
        if (bootstrapped.add(engine)) persist()
    }

    private fun persist() = store.save(nets.mapValues { it.value.toMap() }, bootstrapped.toSet())
}
